{,x-pack/}metricbeat/module/prometheus/{collector,remote_write}: Add `metrics_count` to Prometheus module if `metrics_count` is `true`

shmsr opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## Proposed commit message

Add `metrics_count` metric (counter) to each document (if enabled) for `collector` and `remote_write` datastreams. Please see the sample docs for more info.

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Sample output

<details>
  <summary>Collector</summary>

```json
{
  "_index": ".ds-metricbeat-8.16.0-2024.08.01-000001",
  "_id": "i14sKZEBcmwqWRmK-rWQ",
  "_version": 1,
  "_score": 0,
  "_source": {
    "@timestamp": "2024-08-06T19:31:39.257Z",
    "prometheus": {
      "labels": {
        "instance": "localhost:8081",
        "job": "prometheus"
      },
      "system_memory_usage_bytes": {
        "value": 7351702269.7480135
      },
      "up": {
        "value": 1
      }
    },
    "metrics_count": 2,
    "metricset": {
      "name": "collector",
      "period": 10000
    },
    "service": {
      "type": "prometheus",
      "address": "http://localhost:8081/metrics"
    },
    "event": {
      "dataset": "prometheus.collector",
      "module": "prometheus",
      "duration": 2130125
    },
    "agent": {
      "name": "Subhams-MacBook-Pro.local",
      "type": "metricbeat",
      "version": "8.16.0",
      "ephemeral_id": "6d63b998-1566-4299-aa59-ced8c7a0b7e3",
      "id": "00e1e439-0173-4fb3-8d90-74f4b06ea202"
    },
    "ecs": {
      "version": "8.0.0"
    },
    "host": {
      "mac": [
        "36-1C-20-59-8C-80",
        "36-1C-20-59-8C-84",
        "36-1C-20-59-8C-88",
        "6A-03-C5-9F-E3-B5",
        "6A-03-C5-9F-E3-B6",
        "6A-03-C5-9F-E3-B7",
        "6A-03-C5-9F-E3-D5",
        "6A-03-C5-9F-E3-D6",
        "6A-03-C5-9F-E3-D7",
        "AA-B7-74-48-81-60",
        "BC-D0-74-0D-6F-A5",
        "BE-D0-74-0D-6F-A5",
        "BE-D0-74-D0-53-64",
        "CA-59-C1-CF-D2-0F"
      ],
      "hostname": "Subhams-MacBook-Pro.local",
      "name": "subhams-macbook-pro.local",
      "architecture": "arm64",
      "os": {
        "type": "macos",
        "platform": "darwin",
        "version": "13.6",
        "family": "darwin",
        "name": "macOS",
        "kernel": "22.6.0",
        "build": "22G120"
      },
      "id": "C2677768-AA3E-5023-BD42-EB78BC3453A2",
      "ip": [
        "fe80::6803:c5ff:fe9f:e3d7",
        "fe80::6803:c5ff:fe9f:e3d5",
        "fe80::6803:c5ff:fe9f:e3d6",
        "fe80::bcd0:74ff:fe0d:6fa5",
        "fe80::8e1:bdcb:6244:6bc2",
        "2405:201:d02a:1114:1042:40ca:c385:f622",
        "2405:201:d02a:1114:da8:4a8:4c93:310f",
        "192.168.29.202",
        "fe80::c859:c1ff:fecf:d20f",
        "fe80::c859:c1ff:fecf:d20f",
        "fe80::c646:a13d:9b18:7919",
        "fe80::db4f:7fe7:4b9:cc51",
        "fe80::ce81:b1c:bd2c:69e",
        "fe80::8c9c:d4c3:99a0:8bf8",
        "fe80::781e:7cd2:79db:db56",
        "fe80::2376:e283:a58f:4f88",
        "fe80::7ff1:4c06:56e6:d8e2",
        "192.168.65.1",
        "fe80::bcd0:74ff:fed0:5364",
        "fda2:9011:41d4:6fd3:14bc:39e2:c7ad:76c4"
      ]
    }
  },
  "fields": {
    "beats_state.state.host.hostname": [
      "Subhams-MacBook-Pro.local"
    ],
    "host.os.name.text": [
      "macOS"
    ],
    "host.hostname": [
      "Subhams-MacBook-Pro.local"
    ],
    "host.mac": [
      "36-1C-20-59-8C-80",
      "36-1C-20-59-8C-84",
      "36-1C-20-59-8C-88",
      "6A-03-C5-9F-E3-B5",
      "6A-03-C5-9F-E3-B6",
      "6A-03-C5-9F-E3-B7",
      "6A-03-C5-9F-E3-D5",
      "6A-03-C5-9F-E3-D6",
      "6A-03-C5-9F-E3-D7",
      "AA-B7-74-48-81-60",
      "BC-D0-74-0D-6F-A5",
      "BE-D0-74-0D-6F-A5",
      "BE-D0-74-D0-53-64",
      "CA-59-C1-CF-D2-0F"
    ],
    "host.os.build": [
      "22G120"
    ],
    "service.type": [
      "prometheus"
    ],
    "host.ip": [
      "fe80::6803:c5ff:fe9f:e3d7",
      "fe80::6803:c5ff:fe9f:e3d5",
      "fe80::6803:c5ff:fe9f:e3d6",
      "fe80::bcd0:74ff:fe0d:6fa5",
      "fe80::8e1:bdcb:6244:6bc2",
      "2405:201:d02a:1114:1042:40ca:c385:f622",
      "2405:201:d02a:1114:da8:4a8:4c93:310f",
      "192.168.29.202",
      "fe80::c859:c1ff:fecf:d20f",
      "fe80::c859:c1ff:fecf:d20f",
      "fe80::c646:a13d:9b18:7919",
      "fe80::db4f:7fe7:4b9:cc51",
      "fe80::ce81:b1c:bd2c:69e",
      "fe80::8c9c:d4c3:99a0:8bf8",
      "fe80::781e:7cd2:79db:db56",
      "fe80::2376:e283:a58f:4f88",
      "fe80::7ff1:4c06:56e6:d8e2",
      "192.168.65.1",
      "fe80::bcd0:74ff:fed0:5364",
      "fda2:9011:41d4:6fd3:14bc:39e2:c7ad:76c4"
    ],
    "agent.type": [
      "metricbeat"
    ],
    "logstash_stats.timestamp": [
      "2024-08-06T19:31:39.257Z"
    ],
    "event.module": [
      "prometheus"
    ],
    "host.os.version": [
      "13.6"
    ],
    "host.os.kernel": [
      "22.6.0"
    ],
    "host.os.name": [
      "macOS"
    ],
    "metrics_count": [
      2
    ],
    "agent.name": [
      "Subhams-MacBook-Pro.local"
    ],
    "host.name": [
      "subhams-macbook-pro.local"
    ],
    "beats_state.timestamp": [
      "2024-08-06T19:31:39.257Z"
    ],
    "host.id": [
      "C2677768-AA3E-5023-BD42-EB78BC3453A2"
    ],
    "beats_state.state.host.name": [
      "subhams-macbook-pro.local"
    ],
    "timestamp": [
      "2024-08-06T19:31:39.257Z"
    ],
    "prometheus.system_memory_usage_bytes.value": [
      7351702269.7480135
    ],
    "prometheus.up.value": [
      1
    ],
    "host.os.type": [
      "macos"
    ],
    "kibana_stats.timestamp": [
      "2024-08-06T19:31:39.257Z"
    ],
    "metricset.period": [
      10000
    ],
    "agent.hostname": [
      "Subhams-MacBook-Pro.local"
    ],
    "event.duration": [
      2130125
    ],
    "host.architecture": [
      "arm64"
    ],
    "metricset.name": [
      "collector"
    ],
    "prometheus.labels.instance": [
      "localhost:8081"
    ],
    "@timestamp": [
      "2024-08-06T19:31:39.257Z"
    ],
    "agent.id": [
      "00e1e439-0173-4fb3-8d90-74f4b06ea202"
    ],
    "ecs.version": [
      "8.0.0"
    ],
    "host.os.platform": [
      "darwin"
    ],
    "service.address": [
      "http://localhost:8081/metrics"
    ],
    "agent.ephemeral_id": [
      "6d63b998-1566-4299-aa59-ced8c7a0b7e3"
    ],
    "prometheus.labels.job": [
      "prometheus"
    ],
    "agent.version": [
      "8.16.0"
    ],
    "beats_state.state.host.architecture": [
      "arm64"
    ],
    "host.os.family": [
      "darwin"
    ],
    "event.dataset": [
      "prometheus.collector"
    ]
  }
}
```

</details>


<details>
  <summary>Remote Write</summary>

```json
{
  "@timestamp": [
    "2024-08-06T20:11:35.881Z"
  ],
  "agent.ephemeral_id": [
    "4b5d7b27-4dec-4d60-89c5-5214e9805929"
  ],
  "agent.hostname": [
    "Subhams-MacBook-Pro.local"
  ],
  "agent.id": [
    "00e1e439-0173-4fb3-8d90-74f4b06ea202"
  ],
  "agent.name": [
    "Subhams-MacBook-Pro.local"
  ],
  "agent.type": [
    "metricbeat"
  ],
  "agent.version": [
    "8.16.0"
  ],
  "beats_state.state.host.architecture": [
    "arm64"
  ],
  "beats_state.state.host.hostname": [
    "Subhams-MacBook-Pro.local"
  ],
  "beats_state.state.host.name": [
    "subhams-macbook-pro.local"
  ],
  "beats_state.timestamp": [
    "2024-08-06T20:11:35.881Z"
  ],
  "ecs.version": [
    "8.0.0"
  ],
  "event.dataset": [
    "prometheus.remote_write"
  ],
  "event.module": [
    "prometheus"
  ],
  "host.architecture": [
    "arm64"
  ],
  "host.hostname": [
    "Subhams-MacBook-Pro.local"
  ],
  "host.id": [
    "C2677768-AA3E-5023-BD42-EB78BC3453A2"
  ],
  "host.ip": [
    "fe80::6803:c5ff:fe9f:e3d7",
    "fe80::6803:c5ff:fe9f:e3d5",
    "fe80::6803:c5ff:fe9f:e3d6",
    "fe80::bcd0:74ff:fe0d:6fa5",
    "fe80::8e1:bdcb:6244:6bc2",
    "2405:201:d02a:1114:1042:40ca:c385:f622",
    "2405:201:d02a:1114:da8:4a8:4c93:310f",
    "192.168.29.202",
    "fe80::c859:c1ff:fecf:d20f",
    "fe80::c859:c1ff:fecf:d20f",
    "fe80::c646:a13d:9b18:7919",
    "fe80::db4f:7fe7:4b9:cc51",
    "fe80::ce81:b1c:bd2c:69e",
    "fe80::8c9c:d4c3:99a0:8bf8",
    "fe80::781e:7cd2:79db:db56",
    "fe80::2376:e283:a58f:4f88",
    "fe80::7ff1:4c06:56e6:d8e2",
    "192.168.65.1",
    "fe80::bcd0:74ff:fed0:5364",
    "fda2:9011:41d4:6fd3:14bc:39e2:c7ad:76c4"
  ],
  "host.mac": [
    "36-1C-20-59-8C-80",
    "36-1C-20-59-8C-84",
    "36-1C-20-59-8C-88",
    "6A-03-C5-9F-E3-B5",
    "6A-03-C5-9F-E3-B6",
    "6A-03-C5-9F-E3-B7",
    "6A-03-C5-9F-E3-D5",
    "6A-03-C5-9F-E3-D6",
    "6A-03-C5-9F-E3-D7",
    "AA-B7-74-48-81-60",
    "BC-D0-74-0D-6F-A5",
    "BE-D0-74-0D-6F-A5",
    "BE-D0-74-D0-53-64",
    "CA-59-C1-CF-D2-0F"
  ],
  "host.name": [
    "subhams-macbook-pro.local"
  ],
  "host.os.build": [
    "22G120"
  ],
  "host.os.family": [
    "darwin"
  ],
  "host.os.kernel": [
    "22.6.0"
  ],
  "host.os.name": [
    "macOS"
  ],
  "host.os.name.text": [
    "macOS"
  ],
  "host.os.platform": [
    "darwin"
  ],
  "host.os.type": [
    "macos"
  ],
  "host.os.version": [
    "13.6"
  ],
  "kibana_stats.timestamp": [
    "2024-08-06T20:11:35.881Z"
  ],
  "logstash_stats.timestamp": [
    "2024-08-06T20:11:35.881Z"
  ],
  "metrics_count": [
    6
  ],
  "metricset.name": [
    "remote_write"
  ],
  "prometheus.labels.instance": [
    "localhost:8081"
  ],
  "prometheus.labels.job": [
    "prometheus"
  ],
  "prometheus.scrape_duration_seconds.value": [
    0.0047865
  ],
  "prometheus.scrape_samples_post_metric_relabeling.value": [
    17
  ],
  "prometheus.scrape_samples_scraped.value": [
    17
  ],
  "prometheus.scrape_series_added.value": [
    0
  ],
  "prometheus.system_memory_usage_bytes.value": [
    2583889525.1153283
  ],
  "prometheus.up.value": [
    1
  ],
  "service.type": [
    "prometheus"
  ],
  "timestamp": [
    "2024-08-06T20:11:35.881Z"
  ],
  "_id": "-l5RKZEBcmwqWRmKn-Xc",
  "_index": ".ds-metricbeat-8.16.0-2024.08.01-000001",
  "_score": null
}
```

</details>

shmsr commented:
/test

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane (Team:Elastic-Agent-Data-Plane)

shmsr commented:
@tetianakravchenko / @gizas Could you please look at the changes as well? For more details about the change, I can share an internal document explaining why this change is being made.

belimawr commented on metricbeat/module/prometheus/remote_write/data.go:
```diff
@@ -27,17 +27,20 @@ import (
 
 func (p *remoteWriteEventGenerator) GenerateEvents(metrics model.Samples) map[string]mb.Event {
 	eventList := map[string]mb.Event{}
+	metricCounter := make(map[string]int64, len(metrics)/2)
```
[Question]
Why is this initialised with a length of `len(metrics)/2` instead of `len(metrics)`?

shmsr replied:
In this function, every document is grouped by labels. Let's consider a few Prometheus metrics:

> format: <metric_name><label1=value1,label2=value2,...>

- `abc[k1=v1,k2=v2]`
- `def[k1=v1,k2=v2]`
- `ghi[k3=v3]`
- `xyz[k1=v1]`

So they are grouped as: 

- `abc[k1=v1,k2=v2], def[k1=v1,k2=v2]`
- `ghi[k3=v3]`
- `xyz[k1=v1]`

So, in this case, len(metrics) is 4. But the unique combination of labels is 3. As we are counting the number of metrics per document, we are making a map `metricCounter` that can hold a counter value for each unique label key-val pair (because ultimately that is how many documents will get created). But do not have the number of how many unique combinations are there; so you can say `len(metrics)/2` is just a magic number that we use to pre-alloc the map. It is very very rare that we will have metrics where each and every metric has unique label key-val.

So, `len(metrics)/2` is right in the middle; not essentially the perfect prealloc number but a good enough number and definitely a much better number than `len(metrics)`. But if you have a better magic number suggestion; happy to hear.

mauri870 replied:
Thanks for the explanation. It wasn’t immediately clear to me why this length was chosen; perhaps a brief comment explaining it would be helpful?

shmsr replied:
Hi @mauri870! If you look at the grouping strategy, we're preallocating the map with `len(metrics)/2`, considering that for `n` metrics, `n/2` groups are a reasonable estimate. The grouping is based on unique label key-value pairs representing individual documents. Here's a breakdown:

- We have `n` total metrics
- We preallocate the map with `n/2`
- Each entry in the map holds the count of metrics per document
  - The key is a unique label key-value pair
  - The value is the count of metrics with that exact label combination
- After grouping, len(map) equals len(docs), as each map entry represents a unique document

Maybe `n/2` is not the best number, but I didn't spend too much time but felt choosing `n/2` is quite balanced and works well for average cases.

The optimal capacity depends on the cardinality of your metrics, which can vary widely between different Prometheus setups:

- If you have many metrics with unique label combinations, `n/2` might be too low, leading to reallocations (which is fine; with our balanced `n/2` prealloc at least we reduced some)
- If you have few unique combinations, `n/2` might waste some memory.

mauri870 replied:
Hey, sorry if my previous comment wasn't clear. I appreciate you taking the time to explain the decision; I understand and agree with it. While reading the code, I recognized it as an optimization, but the reasoning behind choosing len(metrics)/2 wasn't immediately clear. I imagine others reviewing the code, perhaps a few months or even a year from now, might have the same question. It could be helpful to add a brief comment to the code explaining why this length was chosen, as there is a valid rationale behind it.

ishleenk17 commented:
@shmsr : Can we please update the data.json file (in both collector and remote_write)

ishleenk17 commented on metricbeat/module/prometheus/_meta/fields.yml:
```diff
@@ -6,6 +6,10 @@
+    - name: metrics_count
+      type: long
+      description: >
+        Count the number of metrics
```
```suggestion
        Count the number of metrics per Elasticsearch document
```

devamanv commented on x-pack/metricbeat/module/prometheus/remote_write/remote_write_test.go:
```diff
@@ -18,6 +18,70 @@ import (
 	"github.com/elastic/elastic-agent-libs/mapstr"
 )
 
+func BenchmarkGenerateEvents(b *testing.B) {
```
Nice to see Benchmarks being added 👍🏼

devamanv commented on metricbeat/module/prometheus/remote_write/data.go:
```diff
@@ -27,18 +27,24 @@ import (
 // DefaultRemoteWriteEventsGeneratorFactory returns the default prometheus events generator
-func DefaultRemoteWriteEventsGeneratorFactory(ms mb.BaseMetricSet) (RemoteWriteEventsGenerator, error) {
-	return &remoteWriteEventGenerator{}, nil
+func DefaultRemoteWriteEventsGeneratorFactory(ms mb.BaseMetricSet, countMetrics bool) (RemoteWriteEventsGenerator, error) {
```
Or you could use the Functional Option pattern like as follows:

``` go
type RemoteWriteEventsGeneratorOption func(*remoteWriteEventGenerator)

func WithCountMetrics(countMetrics bool) RemoteWriteEventsGeneratorOption {
    return func(r *remoteWriteEventGenerator) {
        r.countMetrics = countMetrics
    }
}

func DefaultRemoteWriteEventsGeneratorFactory(ms mb.BaseMetricSet, opts ...RemoteWriteEventsGeneratorOption) (RemoteWriteEventsGenerator, error) {
    generator := &remoteWriteEventGenerator{}
    for _, opt := range opts {
        opt(generator)
    }
    return generator, nil
}

```
And then use it like:

`generator, err := DefaultRemoteWriteEventsGeneratorFactory(ms, WithCountMetrics(true))`

shmsr replied:
This is a very good point. I'll do this.

shmsr replied:
Okay, Aman I tried it. Although adding in metricbeat module is fine, here the complicated part is the x-pack metricbeat. With the FO pattern, not able to get the flow correct in x-pack part of remote write.

While working on this, came across more ways to make it simpler and better (even without FO pattern) but that too did not work. So, MetricSetBuilder is the place where we can do it pretty easily if we'd have defined both the struct types that implement `PromEventsGenerator` in metricbeat module and not separately i.e., different in metricbeat and x-pack metricbeat. Cannot even import the type as it would create cyclic dependency A calls B, B calls A.

If the remote_write code was written similarly to that of collector, then also it could've been possible. I hope I didn't miss anything, let me know if you still think there is a way to add FO pattern in a simple, clean way.

devamanv replied:
@shmsr I tried the changes I suggested. I was able to get it to work using the FO pattern. Here's the diff patch of the changes for you to refer to.  See if these changes look good and can be incorporated. 

```diff
diff --git a/metricbeat/module/prometheus/remote_write/data.go b/metricbeat/module/prometheus/remote_write/data.go
index 6fd57358bb..fa2b4796b2 100644
--- a/metricbeat/module/prometheus/remote_write/data.go
+++ b/metricbeat/module/prometheus/remote_write/data.go
@@ -27,18 +27,30 @@ import (
 )
 
 // DefaultRemoteWriteEventsGeneratorFactory returns the default prometheus events generator
-func DefaultRemoteWriteEventsGeneratorFactory(ms mb.BaseMetricSet, metricsCount bool) (RemoteWriteEventsGenerator, error) {
-	return &remoteWriteEventGenerator{metricsCount: metricsCount}, nil
+func DefaultRemoteWriteEventsGeneratorFactory(ms mb.BaseMetricSet, opts ...RemoteWriteEventsGeneratorOption) (RemoteWriteEventsGenerator, error) {
+	generator := &RemoteWriteEventGenerator{}
+	for _, opt := range opts {
+		opt(generator)
+	}
+
+	return generator, nil
 }
 
-type remoteWriteEventGenerator struct {
+type RemoteWriteEventGenerator struct {
 	metricsCount bool
 }
 
-func (p *remoteWriteEventGenerator) Start() {}
-func (p *remoteWriteEventGenerator) Stop()  {}
+type RemoteWriteEventsGeneratorOption func(r *RemoteWriteEventGenerator)
+
+func WithCountMetrics(countMetrics bool) RemoteWriteEventsGeneratorOption {
+	return func(r *RemoteWriteEventGenerator) {
+		r.metricsCount = countMetrics
+	}
+}
+func (p *RemoteWriteEventGenerator) Start() {}
+func (p *RemoteWriteEventGenerator) Stop()  {}
 
-func (p *remoteWriteEventGenerator) GenerateEvents(metrics model.Samples) map[string]mb.Event {
+func (p *RemoteWriteEventGenerator) GenerateEvents(metrics model.Samples) map[string]mb.Event {
 	eventList := map[string]mb.Event{}
 
 	for _, metric := range metrics {
diff --git a/metricbeat/module/prometheus/remote_write/remote_write.go b/metricbeat/module/prometheus/remote_write/remote_write.go
index 31d119e9d0..95048f0526 100644
--- a/metricbeat/module/prometheus/remote_write/remote_write.go
+++ b/metricbeat/module/prometheus/remote_write/remote_write.go
@@ -52,7 +52,7 @@ type RemoteWriteEventsGenerator interface {
 }
 
 // RemoteWriteEventsGeneratorFactory creates a RemoteWriteEventsGenerator when instanciating a metricset
-type RemoteWriteEventsGeneratorFactory func(ms mb.BaseMetricSet, metricsCount bool) (RemoteWriteEventsGenerator, error)
+type RemoteWriteEventsGeneratorFactory func(ms mb.BaseMetricSet, opts ...RemoteWriteEventsGeneratorOption) (RemoteWriteEventsGenerator, error)
 
 type MetricSet struct {
 	mb.BaseMetricSet
@@ -69,7 +69,7 @@ func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
 		return nil, err
 	}
 
-	promEventsGen, err := DefaultRemoteWriteEventsGeneratorFactory(base, config.MetricsCount)
+	promEventsGen, err := DefaultRemoteWriteEventsGeneratorFactory(base, WithCountMetrics(config.MetricsCount))
 	if err != nil {
 		return nil, err
 	}
@@ -99,7 +99,7 @@ func MetricSetBuilder(genFactory RemoteWriteEventsGeneratorFactory) func(base mb
 			return nil, err
 		}
 
-		promEventsGen, err := genFactory(base, config.MetricsCount)
+		promEventsGen, err := genFactory(base, WithCountMetrics(config.MetricsCount))
 		if err != nil {
 			return nil, err
 		}
diff --git a/metricbeat/module/prometheus/remote_write/remote_write_test.go b/metricbeat/module/prometheus/remote_write/remote_write_test.go
index f6f7b03eda..fc00754997 100644
--- a/metricbeat/module/prometheus/remote_write/remote_write_test.go
+++ b/metricbeat/module/prometheus/remote_write/remote_write_test.go
@@ -28,7 +28,7 @@ import (
 
 // TestGenerateEventsCounter tests counter simple cases
 func TestGenerateEventsCounter(t *testing.T) {
-	g := remoteWriteEventGenerator{}
+	g := RemoteWriteEventGenerator{}
 
 	timestamp := model.Time(424242)
 	timestamp1 := model.Time(424243)
@@ -237,7 +237,7 @@ func TestMetricsCount(t *testing.T) {
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			generator := remoteWriteEventGenerator{
+			generator := RemoteWriteEventGenerator{
 				metricsCount: true,
 			}
 
diff --git a/x-pack/metricbeat/module/prometheus/remote_write/config.go b/x-pack/metricbeat/module/prometheus/remote_write/config.go
index 9e86facb1d..0f0d91ead7 100644
--- a/x-pack/metricbeat/module/prometheus/remote_write/config.go
+++ b/x-pack/metricbeat/module/prometheus/remote_write/config.go
@@ -10,6 +10,7 @@ import (
 )
 
 type config struct {
+	MetricsCount  bool          `config:"metrics_count"`
 	UseTypes      bool          `config:"use_types"`
 	RateCounters  bool          `config:"rate_counters"`
 	TypesPatterns TypesPatterns `config:"types_patterns" yaml:"types_patterns,omitempty"`
diff --git a/x-pack/metricbeat/module/prometheus/remote_write/data.go b/x-pack/metricbeat/module/prometheus/remote_write/data.go
index bf91a10a70..4c946da5e7 100644
--- a/x-pack/metricbeat/module/prometheus/remote_write/data.go
+++ b/x-pack/metricbeat/module/prometheus/remote_write/data.go
@@ -36,7 +36,7 @@ type histogram struct {
 	metricName string
 }
 
-func remoteWriteEventsGeneratorFactory(base mb.BaseMetricSet, metricsCount bool) (rw.RemoteWriteEventsGenerator, error) {
+func remoteWriteEventsGeneratorFactory(base mb.BaseMetricSet, opts ...rw.RemoteWriteEventsGeneratorOption) (rw.RemoteWriteEventsGenerator, error) {
 	config := defaultConfig
 	if err := base.Module().UnpackConfig(&config); err != nil {
 		return nil, err
@@ -51,7 +51,7 @@ func remoteWriteEventsGeneratorFactory(base mb.BaseMetricSet, metricsCount bool)
 		g := remoteWriteTypedGenerator{
 			counterCache: counters,
 			rateCounters: config.RateCounters,
-			metricsCount: metricsCount,
+			metricsCount: config.MetricsCount,
 		}
 
 		var err error
@@ -67,7 +67,7 @@ func remoteWriteEventsGeneratorFactory(base mb.BaseMetricSet, metricsCount bool)
 		return &g, nil
 	}
 
-	return rw.DefaultRemoteWriteEventsGeneratorFactory(base, metricsCount)
+	return rw.DefaultRemoteWriteEventsGeneratorFactory(base, opts...)
 }
```
**_Note: These changes are just to demonstrate what I suggested in the comment up above. If you think these changes break the functionality, feel free to ignore them in favour of correctness._**

shmsr replied:
Yes, thank you!! This helps! I'll apply this patch!

shmsr replied:
Done. Please check!

shmsr commented on metricbeat/module/prometheus/_meta/fields.yml:
```diff
@@ -6,6 +6,10 @@
+    - name: metrics_count
+      type: long
+      description: >
+        Count the number of metrics
```
Makes sense.

shmsr replied:
Done!

shmsr commented on x-pack/metricbeat/module/prometheus/remote_write/data.go:
```diff
@@ -66,10 +67,11 @@ func remoteWriteEventsGeneratorFactory(base mb.BaseMetricSet) (remote_write.Remo
 	}
 
-	return remote_write.DefaultRemoteWriteEventsGeneratorFactory(base)
+	return remote_write.DefaultRemoteWriteEventsGeneratorFactory(base, countMetrics)
```
Right.

shmsr replied:
Done!

gizas commented on x-pack/metricbeat/module/prometheus/remote_write/remote_write_test.go:
```diff
@@ -18,6 +18,70 @@ import (
 	"github.com/elastic/elastic-agent-libs/mapstr"
 )
 
+func BenchmarkGenerateEvents(b *testing.B) {
```
Can we consider writting  an integration test for the metrics_count enable?

You can create only one for the collector and add a comment that same stands for remote_write.
Idea would be to validate the metrics_count for one or two event families/ labels?

See eg from histograms: https://github.com/elastic/beats/blob/d732d5d5d211eb4bf39ce2b6f5689128dee016b8/x-pack/metricbeat/module/prometheus/collector/histogram_test.go

shmsr replied:
@gizas Here's the sample output when I tested the change with actual Prometheus stats: https://github.com/elastic/beats/pull/40411#issue-2441684623 (See: **Sample output**)

shmsr replied:
> Can we consider writting an integration test for the metrics_count enable as well?

Sure.

shmsr replied:
Test added in this commit: https://github.com/elastic/beats/pull/40411/commits/3d93923d7f4d93356ea495ee3de4ac17c3db0f12

gizas replied:
(Sorry still have not run the tests locally, will try to do it and let u know)

shmsr replied:
> My comment was more to compare those metrics with actual prometheus visualisations because our users this is what will do. They will take our metrics_count and will try to compare with an actual prometheus visualisation
>
> For eg. in remote_write we had (see old https://github.com/elastic/beats/issues/17079#issuecomment-612878489) we had rate(prometheus_remote_storage_samples_total[2m]) . Will this number be close to the sum (metrics_count) for period of 2m in our Kibana?
>
> I recall prometheus_tsdb_head_series shows the number of active time series.
>
> As long as you have a local setup I think a comparison would be benefitial even for us to answer following problems

Sure, will check this. Btw, I'll also DM you the doc where we have written why we want the field.

shmsr replied:
I did test this yesterday. I am not able to see `prometheus_tsdb_head_series`. I checked Prometheus' Server UI and also checked if it is sending this to remote_write destination i.e., metricbeat; but no. But I did see `prometheus.scrape_samples_scraped.value` in Kibana and a few more. But nothing with *tsdb* or *remote_write*.

shmsr commented on metricbeat/module/prometheus/_meta/fields.yml:
```diff
@@ -6,6 +6,10 @@
   release: ga
   settings: ["ssl", "http"]
   fields:
+    - name: metrics_count
```
Yeah, I know. The reason behind the same is internal there was a proposal written before the implementation and there we mentioned that we would add a field called `metrics_count`.

But while implementing, `count_metrics`/`countMetrics` made more sense as a config and a parameter name. But I agree this is a valid point.


Should I make this change:

- s/count_metrics/metrics_count
- s/countMetrics/metricsCount

Doesn't change much; we can go ahead with this.

shmsr replied:
Done!

shmsr commented:
After running "go test -data -run TestData .", I understand that the PR changes might be a little hard to review given the changes in `testdata`; lots of files.

Here's something that might help you while reviewing: https://github.com/elastic/integrations/issues/8028#issuecomment-1765164454 (just change the `data-path` with the pattern (e.g., "expected.json" you want to hide). Save the bookmark and when you are reviewing; click on the bookmark. It might help!

shmsr commented on metricbeat/module/prometheus/collector/collector_test.go:
```diff
@@ -380,3 +384,140 @@ func TestSkipMetricFamily(t *testing.T) {
+	assert.Equal(t, expectedMetricsCount, metricsCount, "Metrics count does not match expected")
+}
+
+// NOTE(shmsr): If "test_histogram_bucket{environment="prod",service="api",le="0.1"} 0"
```
Spent debugging this issue for hours on why tests are passing on the local machine (MacOS M1) but not on CI. I've written the details here. Please read it carefully; maybe the parser needs to be fixed. There's already a related issue opened for this one which I have linked.

cc: @gizas @devamanv

shmsr replied:
Yep, exactly! I've already linked this in the comment itself of this particular test: https://github.com/elastic/beats/blob/42df5573e18da20d07f43123fd7d6a0ca00c2254/metricbeat/module/prometheus/collector/collector_test.go#L476

We have to fix the NaN handling.

mauri870 replied:
Oh, thanks, I missed that. Glad the issue is well known and we are tracking it already.

shmsr commented:
To the reviewers,

I have updated the code and also rewrote some tests. There were issues with the metric count for certain cases of histograms and summaries, particularly in the collector where PromEvent may contain 2 events. Previously, we were only counting 1 event instead of 2 due to the earlier logic. Now, I have written unit tests for all scenarios.

Also found a pesky bug that is platform dependent. See: https://github.com/elastic/beats/pull/40411/files#r1740184901
Also, a tip: https://github.com/elastic/beats/pull/40411#issuecomment-2316220724

ishleenk17 commented on metricbeat/module/prometheus/remote_write/data.go:
```diff
@@ -86,18 +82,18 @@ func (p *remoteWriteEventGenerator) GenerateEvents(metrics model.Samples) map[st
+		for _, e := range eventList {
 			// In x-pack prometheus module, the metrics are nested under the "prometheus" key directly.
 			// whereas in non-x-pack prometheus module, the metrics are nested under the "prometheus.metrics" key.
 			// Also, it is important that we do not just increment by 1 for each e.ModuleFields["metrics"] may have more than 1 metric.
```
This is to accomodate the histogram/summary scenario?

shmsr replied:
Yes!

shmsr commented:
@gizas Did you get a chance to look at this PR again?

devamanv commented on x-pack/metricbeat/module/prometheus/remote_write/data.go:
```diff
@@ -182,12 +186,31 @@ func (g remoteWriteTypedGenerator) GenerateEvents(metrics model.Samples) map[str
+			// we subtract 1 in case the e.ModuleFields["labels"] also exists.
+			//
+			// See unit tests for the same.
+			_, ok := e.ModuleFields["labels"]
+			if ok {
+				e.RootFields["metrics_count"] = len(e.ModuleFields) - 1
+			} else {
+				e.RootFields["metrics_count"] = len(e.ModuleFields)
+			}
```
I think it will be more intuitive to name the flags that actually denotes the meaning of the flag. 
```suggestion
			if _, hasLabels := e.ModuleFields["labels"]; hasLabels {
				e.RootFields["metrics_count"] = len(e.ModuleFields) - 1
				continue
			} 	
			e.RootFields["metrics_count"] = len(e.ModuleFields)
```

shmsr replied:
Cool. s/ok/hasLabels in related pieces of code.

Also for consistency across the setting of metrics_count related changes in collector and remote_write, using if-else (not if along with continue) pattern because here it works fine but a similar pattern with `continue` in collector won't work because the event is being reported in the very same loop. So, keeping it if-else pattern; this is also idiomatic and it's pretty clear too.

shmsr replied:
Changes: https://github.com/elastic/beats/pull/40411/commits/b1408a73521d639ca80f90b7362b2a63e778829f

shmsr commented on x-pack/metricbeat/module/prometheus/remote_write/data.go:
```diff
@@ -130,6 +132,7 @@ func (g remoteWriteTypedGenerator) GenerateEvents(metrics model.Samples) map[str
 		// join metrics with same labels in a single event
 		if _, ok := eventList[labelsHash]; !ok {
 			eventList[labelsHash] = mb.Event{
+				RootFields:   make(mapstr.M),
```
Resolved.

devamanv commented on metricbeat/module/prometheus/remote_write/remote_write_test.go:
```diff
@@ -78,3 +78,188 @@ func TestGenerateEventsCounter(t *testing.T) {
+				count, ok := event.RootFields["metrics_count"]
+				assert.True(t, ok, "metrics_count should be present")
+
+				labels, ok := event.ModuleFields["labels"].(mapstr.M)
```
I noticed that we are using  `assert.True(t, ok, "labels should be present")` in the x-pack counterpart. Any reason that we shouldn't be doing the same thing here?

shmsr replied:
No. Both should be similar. I'll make them same. Changing in x-pack.

shmsr replied:
Change: https://github.com/elastic/beats/pull/40411/commits/274f104f49e177072726ef20410442be5f43be78

devamanv commented on x-pack/metricbeat/metricbeat.reference.yml:
```diff
@@ -1325,6 +1325,9 @@ metricbeat.modules:
   #  - /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
 
+  # Count number of metrics present in Elasticsearch document (default: false)
+  #metrics_count: false
```
We have already specified that the default value is `false`. Should we make `true` instead? Like we do for all other fields with default values?

shmsr replied:
No strong opinion on this as this is a reference file and also a metricbeat config. If we are doing it for other fields then it makes sense.

I'll make the change.

devamanv commented on metricbeat/module/prometheus/collector/collector.go:
```diff
@@ -177,9 +189,48 @@ func (m *MetricSet) Fetch(reporter mb.ReporterV2) error {
+			// So, len(v) will be 2 in the above example.
+			// Similarly, it will happen for x-pack prometheus module too. Please see
+			// the unit tests for the same.
+			switch m.xPack {
+			case true:
+				// As, metrics are nested under the "prometheus" key in case of x-pack,
+				// labels is also nested under the "prometheus" key. So, we need to
+				// make sure we subtract 1 in case the e["labels"]
+				// also exists.
+				if _, hasLabels := e["labels"].(mapstr.M); hasLabels {
+					event.RootFields.Put("metrics_count", len(e)-1)
+				} else {
+					event.RootFields.Put("metrics_count", len(e))
+				}
+			default:
+				if v, ok := e["metrics"].(mapstr.M); ok {
+					event.RootFields.Put("metrics_count", len(v))
+				}
+			}
```
I'm thinking if we need this additional variable xPack at all. Since we have access to the metricset object, we could simply
```suggestion
			switch m.promEventsGen.(type) { {
			case *promEventGenerator: // x-pack
			          if v, ok := e["metrics"].(mapstr.M); ok {
					event.RootFields.Put("metrics_count", len(v))
				 }
			default: // non-xpack
				// As, metrics are nested under the "prometheus" key in case of x-pack,
				// labels is also nested under the "prometheus" key. So, we need to
				// make sure we subtract 1 in case the e["labels"]
				// also exists.
				if _, hasLabels := e["labels"].(mapstr.M); hasLabels {
					event.RootFields.Put("metrics_count", len(e)-1)
				} else {
					event.RootFields.Put("metrics_count", len(e))
				}
			}
```

shmsr replied:
Yes, my first draft change was this only. 

But this may seem unclear to some future readers of the code, so I added the xPack so that it is clear, and they can also use the xPack variable in case they need it.

What do you think?

devamanv replied:
IMO, since you already added a note about the generator type while asserting it, it should be enough the readers to know why it is the way it is. Besides, metricset is accessible from both collector as well as remote write. So, we should be able to use no matter where it's needed. 

That said, I am all in for keeping it this way too. My question was more regarding the use of an extra field, which seemed redundant to me.

shmsr replied:
I did take a look at this and I recall another rationale behind introducing a new boolean variable. The primary goal was to optimize performance by eliminating the need for repeated type-checking in every `Fetch` operation (happens at set intervals). By performing the type assertion once during initialization and storing the result in a boolean field, we avoid the overhead of type-checking in each `Fetch` call. So it replaces frequent type assertions with a simple boolean check.

So keeping it as it is?

devamanv replied:
Sounds good. Resolving the conversation in favour of keeping it unchanged.

ishleenk17 reviewed (approved):
Looks good!

gizas commented:
Thanks for the tests added. LGTM and also help to understand how the metrics_count is calculated
